File: Java/Command/SmartHome/src/com/ericgtkb/SmartHome.java
package com.ericgtkb;

import java.util.HashMap;
import java.util.Map;

// Invoker class
public class SmartHome {
    private Map<String, Command> onCommands;
    private Map<String, Command> offCommands;

    public SmartHome() {
        onCommands = new HashMap<>();
        offCommands = new HashMap<>();
    }

    public void addCommand(String name, Command onCommand, Command offCommand) {
        onCommands.put(name, onCommand);
        offCommands.put(name, offCommand);
    }

    public void turnOn(String name) {
        onCommands.get(name).execute();
    }

    public void turnOff(String name) {
        offCommands.get(name).execute();
    }
}

File: C++/Command/SmartHome/smart_home.cc
#include "smart_home.h"

SmartHome::SmartHome() : on_commands(), off_commands() {}

void SmartHome::add_command(const std::string& name, std::unique_ptr<Command>& on_command,
                            std::unique_ptr<Command>& off_command) {
    on_commands[name] = std::move(on_command);
    off_commands[name] = std::move(off_command);
}

void SmartHome::turn_on(const std::string &name) {
    on_commands[name]->execute();
}

void SmartHome::turn_off(const std::string &name) {
    off_commands[name]->execute();
}

